project/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -101,6 +131,95 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     return elections;
   }
 
+  /**
+   * Takes an {@code Entity} object which represents an election, and returns true if
+   * this election is deemed relevant to the given {@code address}. If the {@code listAll}
+   * parameter is true, then the election is always deemed relevant.
+   */
+  private boolean isRelevantElection(Entity election, String address, boolean listAllElections) {
+    resetRelevantNonspecificFlag();
+    if (listAllElections) {
+      return true;
+    } else {
+      // Query the Civic Information API for whether {@code election} is relevant to
+      // {@code address}.
+      String queryUrl =
+        String.format("%s&address=%s&electionId=%s", VOTER_INFO_QUERY_URL,
+                      address.replace(" ", "%20").replace("\"", "%22").replace(",", "%2C"),
+                      (String) election.getProperty("queryId"));
+      String addressElectionResponse;
+      try {
+        addressElectionResponse = queryCivicInformation(queryUrl);
+      } catch (Exception e) {
+        return false;
+      }
+      return parseResponseForRelevancy(addressElectionResponse);
+    }
+  }
+
+  /** 
+   * Queries the Civic Information API and retrieves JSON response as a String.
+   *
+   * @throws ClientProtocolException if the GET request to the Civic Information API fails.
+   * @throws SocketException if {@code queryUrl} is ill-constructed, such as with a null value
+   *     as the election query ID, causing {@code httpclient.execute(httpGet, responseHandler);}
+   *     to fail.
+   */
+  private String queryCivicInformation(String queryUrl) throws IOException, SocketException {
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    HttpGet httpGet = new HttpGet(queryUrl);
+    return requestHttpAndBuildJsonResponseFromCivicInformation(httpClient, httpGet);
+  }
+
+  /** 
+   * Makes HTTP GET request to the Civic Information API amd returns HTTP JSON response as a
+   * String. This method is given default visibility for testing purposes.
+   *
+   * @throws ClientProtocolException if the GET request to the Civic Information API fails.
+   * @throws SocketException if {@code queryUrl} is ill-constructed, such as with a null value
+   *     as the election query ID, causing {@code httpclient.execute(httpGet, responseHandler);}
+   *     to fail.
+   * @see <a href=
+   *    "https://hc.apache.org/httpcomponents-client-ga/httpclient/examples/org/apache/" +
+   *    "http/examples/client/ClientWithResponseHandler.java">Code reference</a>
+   */
+  String requestHttpAndBuildJsonResponseFromCivicInformation(CloseableHttpClient httpClient,
```

Comment:
nit. Avoid long function names when possible. Rename this to requestAndBuildCivicInformationResponse